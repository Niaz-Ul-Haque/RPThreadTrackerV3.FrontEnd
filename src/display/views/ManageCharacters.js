import React, { Component } from 'react';
import {
	Row, Col
} from 'reactstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import NewCharacterFormCard from '../components/characters/NewCharacterFormCard';
import CurrentCharacterTable from '../components/characters/CurrentCharacterTable';
import { generateRandomThread, fetchUserSettings, setHasDashboardAtAGlanceHidden, fetchActiveThreads, fetchCharacters } from '../../infrastructure/actions';
import { getMyTurnThreads, getTheirTurnThreads, getQueuedThreads, getRecentActivity } from '../../infrastructure/selectors';

const propTypes = {
	characters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	dispatch: PropTypes.func.isRequired
};

function mapStateToProps(state) {
	const {
		characters
	} = state;
	return {
		characters
	};
}

class ManageCharacters extends Component {
	componentDidMount() {
		const { dispatch } = this.props;
		if (!this.props.characters || !this.props.characters.length) {
			dispatch(fetchCharacters());
		}
	}

	render() {
		const {
			characters
		} = this.props;
		return (
			<div className="animated fadeIn dashboard-container">
				<Row>
					<Col>
						<NewCharacterFormCard />
					</Col>
				</Row>
				<Row>
					<Col>
						<CurrentCharacterTable characters={characters} />
					</Col>
				</Row>
			</div >
		);
	}
}

ManageCharacters.propTypes = propTypes;
export default connect(mapStateToProps)(ManageCharacters);
