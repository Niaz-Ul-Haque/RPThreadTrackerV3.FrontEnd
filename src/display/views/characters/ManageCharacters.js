// #region imports
import React, { Component } from 'react';
import { Row, Col } from 'reactstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import CurrentCharacterTable from './components/CurrentCharacterTable';
import { getIsLoadingIconVisible } from '../../../infrastructure/selectors';
import { openUpsertCharacterModal, fetchCharacters, upsertCharacter, openUntrackCharacterModal } from '../../../infrastructure/actions';
// #endregion imports

const propTypes = {
	characters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	isLoadingIconVisible: PropTypes.bool.isRequired,
	fetchCharacters: PropTypes.func.isRequired,
	openUpsertCharacterModal: PropTypes.func.isRequired,
	upsertCharacter: PropTypes.func.isRequired,
	openUntrackCharacterModal: PropTypes.func.isRequired
};

function mapStateToProps(state) {
	const {
		characters
	} = state;
	const isLoadingIconVisible = getIsLoadingIconVisible(state);
	return {
		characters,
		isLoadingIconVisible
	};
}

class ManageCharacters extends Component {
	constructor(props) {
		super(props);
		this.toggleCharacterIsOnHiatus = this.toggleCharacterIsOnHiatus.bind(this);
	}

	componentDidMount() {
		if (!this.props.characters || !this.props.characters.length) {
			this.props.fetchCharacters();
		}
	}

	toggleCharacterIsOnHiatus(character) {
		const updatedCharacter = {
			...character, isOnHiatus: !character.isOnHiatus
		};
		this.props.upsertCharacter(updatedCharacter);
	}

	render() {
		const {
			characters,
			isLoadingIconVisible
		} = this.props;
		return (
			<div className="animated fadeIn dashboard-container">
				<Row>
					<Col>
						<CurrentCharacterTable
							data-spec="manage-characters-current-character-table"
							characters={characters}
							openUpsertCharacterModal={this.props.openUpsertCharacterModal}
							toggleCharacterIsOnHiatus={this.toggleCharacterIsOnHiatus}
							openUntrackCharacterModal={this.props.openUntrackCharacterModal}
							isLoadingIconVisible={isLoadingIconVisible}
						/>
					</Col>
				</Row>
			</div >
		);
	}
}

ManageCharacters.propTypes = propTypes;
export default connect(mapStateToProps, {
	fetchCharacters,
	openUntrackCharacterModal,
	openUpsertCharacterModal,
	upsertCharacter
})(ManageCharacters);
