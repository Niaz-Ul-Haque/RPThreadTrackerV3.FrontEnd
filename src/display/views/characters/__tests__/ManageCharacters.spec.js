// #region imports
import React from 'react';
import { shallowWithState, getSpecWrapper } from '~/testhelpers/helpers.unit';
import ManageCharacters from '../ManageCharacters';
// #endregion imports

// #region mocks
jest.mock('../../../../infrastructure/actions', () => ({}));
jest.mock('../../../../infrastructure/selectors', () => ({
	getIsLoadingIconVisible: () => true,
	getThreadCountsByCharacter: () => ({ 5: 10, 15: 20 })
}));
jest.mock('../components/CurrentCharacterTable', () => 'CurrentCharacterTable');
// #endregion mocks

const createTestProps = propOverrides => ({
	fetchActiveThreads: jest.fn(),
	fetchCharacters: jest.fn(),
	openUpsertCharacterModal: jest.fn(),
	upsertCharacter: jest.fn(),
	openUntrackCharacterModal: jest.fn(),
	...propOverrides
});

const createTestState = stateOverrides => ({
	characters: [
		{ characterId: 1, characterName: 'Character 1' },
		{ characterId: 2, characterName: 'Character 2' }
	],
	activeThreads: [{ threadId: 1, userTitle: 'Active 1' }, { threadId: 2, userTitle: 'Active 2' }, { threadId: 3, userTitle: 'Active 3' }],
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
	});
	describe('content', () => {
		it('should pass props to character table', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const table = getSpecWrapper(element, 'manage-characters-current-character-table');
			expect(table.props().characters).toHaveLength(2);
			expect(table.props().openUpsertCharacterModal).toBeDefined();
			expect(table.props().toggleCharacterIsOnHiatus).toBeDefined();
			expect(table.props().openUntrackCharacterModal).toBeDefined();
			expect(table.props().isLoadingIconVisible).toBe(true);
			expect(table.props().threadCounts).toEqual({ 5: 10, 15: 20 });
		});
	});
});

describe('behavior', () => {
	describe('componentDidMount', () => {
		it('should fetch characters when characters are not loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({ fetchCharacters });
			const state = createTestState({ characters: [] });
			const jsx = (<ManageCharacters {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(1);
		});
		it('should not fetch characters when characters are loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({ fetchCharacters });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(0);
		});
		it('should fetch active threads when active threads are not loaded', () => {
			const fetchActiveThreads = jest.fn();
			const props = createTestProps({ fetchActiveThreads });
			const state = createTestState({ activeThreads: [] });
			const jsx = (<ManageCharacters {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchActiveThreads).toHaveBeenCalledTimes(1);
		});
		it('should not fetch active threads when active threads are loaded', () => {
			const fetchActiveThreads = jest.fn();
			const props = createTestProps({ fetchActiveThreads });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchActiveThreads).toHaveBeenCalledTimes(0);
		});
	});
	describe('openUpsertCharacterModal', () => {
		it('should be triggered when child component prop is triggered', () => {
			const openUpsertCharacterModal = jest.fn();
			const props = createTestProps({ openUpsertCharacterModal });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'manage-characters-current-character-table');
			child.props().openUpsertCharacterModal();
			expect(openUpsertCharacterModal).toHaveBeenCalledTimes(1);
		});
	});
	describe('openUntrackCharacterModal', () => {
		it('should be triggered when child component prop is triggered', () => {
			const openUntrackCharacterModal = jest.fn();
			const props = createTestProps({ openUntrackCharacterModal });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'manage-characters-current-character-table');
			child.props().openUntrackCharacterModal();
			expect(openUntrackCharacterModal).toHaveBeenCalledTimes(1);
		});
	});
	describe('toggleCharacterIsOnHiatus', () => {
		it('should be triggered when child component prop is triggered', () => {
			const upsertCharacter = jest.fn();
			const props = createTestProps({ upsertCharacter });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'manage-characters-current-character-table');
			child.props().toggleCharacterIsOnHiatus({ characterId: 3, isOnHiatus: false });
			expect(upsertCharacter).toHaveBeenCalledTimes(1);
		});
		it('should dispatch update character action with isOnHiatus inverted', () => {
			const upsertCharacter = jest.fn();
			const props = createTestProps({ upsertCharacter });
			const state = createTestState();
			const jsx = (<ManageCharacters {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.instance().toggleCharacterIsOnHiatus({ characterId: 3, isOnHiatus: false });
			expect(upsertCharacter).toHaveBeenCalledWith({ characterId: 3, isOnHiatus: true });
		});
	});
});
