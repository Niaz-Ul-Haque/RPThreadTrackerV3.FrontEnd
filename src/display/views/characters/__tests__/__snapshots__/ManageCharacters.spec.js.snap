// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<div
  className="animated fadeIn dashboard-container"
>
  <Row
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <CurrentCharacterTable
        characters={
          Array [
            Object {
              "characterId": 1,
              "characterName": "Character 1",
            },
            Object {
              "characterId": 2,
              "characterName": "Character 2",
            },
          ]
        }
        data-spec="manage-characters-current-character-table"
        openUntrackCharacterModal={[MockFunction]}
        openUpsertCharacterModal={[MockFunction]}
        toggleCharacterIsOnHiatus={[Function]}
      />
    </Col>
  </Row>
</div>
`;
